### ISSUE TITLE
`std.io.AnyWriter` has very high overhead

### ISSUE THREAD

**Issue by silversquirl** [ziglang/zig#21566]
`std.io.AnyWriter` (and hence `std.io.GenericWriter`, since that uses `AnyWriter` internally) has a large amount of overhead in tight loops, as it hampers inlining.

The following code snippet demonstrates an extreme case of the issue, but it comes up a lot in practice too:

```zig
pub fn main() !void {
    const devnull = try std.fs.openFileAbsolute("/dev/null", .{ .mode = .write_only });
    var timer = try std.time.Timer.start();
    {
        // Write 1MiB to /dev/null, one byte at a time, using std.io.BufferedWriter
        var bufw = std.io.bufferedWriter(devnull.writer());
        const w = bufw.writer();
        for (0..1 << 20) |i| {
            try w.writeByte(@truncate(i));
        }
        try bufw.flush();

        std.debug.print("BufferedWriter: {}\n", .{std.fmt.fmtDuration(timer.read())});
    }

    {
        timer.reset();
        // Write 1MiB to /dev/null, one byte at a time, using a custom buffering solution
        var buf: [4096]u8 = undefined;
        var idx: usize = 0;
        const w = devnull.writer();
        for (0..1 << 20) |i| {
            buf[idx] = @truncate(i);
            idx += 1;

            if (idx >= buf.len) {
                try w.writeAll(&buf);
                idx = 0;
            }
        }
        try w.writeAll(buf[0..idx]);

        std.debug.print("Custom buffering: {}\n", .{std.fmt.fmtDuration(timer.read())});
    }
}

const std = @import("std");
```

On my machine, this code produces the following result: 
```
$ zig run writer_slow.zig -Doptimize=ReleaseFast
BufferedWriter: 35.568ms
Custom buffering: 1.993ms
```

Using a custom buffering solution rather than `std.io.BufferedWriter` results in an almost 20x speedup, simply by avoiding the overhead of `std.io.AnyWriter`.
This pretty much defeats the whole point of having `BufferedWriter` in the first place, since `AnyWriter` destroys the good write performance it would otherwise provide.

**Comment by marler8997** [comment]
Interesting stuff.  I'm getting the same numbers on my macbook air and Windows desktop. I tested the performance on some older releases of zig down to `0.10.0` and found the fastest to be `0.11.0`.  I copied the implementation from `0.11.0` into the example and it takes about 60% of the time compared to the new implementation that goes through `AnyWriter`, still far off from the "no writer" benchmark which is around 8% of the time.  In addition I wrote a theoretical `MetaWriter` which gets rid of `writeFn` altogether and instead queries the underlying `Context` type to call an appropriate function and I got it to about 15% of the runtime (about 2x slower than having no writer).  Here's that code:

```zig
const std = @import("std");

pub fn main() !void {
    const devnull = try std.fs.openFileAbsolute(
        if (@import("builtin").os.tag == .windows) "\\\\.\\NUL" else "/dev/null",
        .{ .mode = .write_only },
    );

    var timer = try std.time.Timer.start();

    for (0 .. 10) |_| {
        timer.reset();
        try testNoWriter(devnull);
        std.debug.print("No Writer: {}\n", .{std.fmt.fmtDuration(timer.read())});
        timer.reset();
        try testBufferedWriter(devnull);
        std.debug.print("BufferedWriter: {}\n", .{std.fmt.fmtDuration(timer.read())});
        timer.reset();
        try testBufferedWriterZig_0_11_0(devnull);
        std.debug.print("BufferedWriter(zig 0.11.0): {}\n", .{std.fmt.fmtDuration(timer.read())});
        timer.reset();
        try testBufferedWriterMeta(devnull);
        std.debug.print("BufferedWriterMeta: {}\n", .{std.fmt.fmtDuration(timer.read())});
    }
}

// Each example writes 1MiB to /dev/null, one byte at a time

// test today's buffered writer which goes through AnyWriter
fn testBufferedWriter(devnull: std.fs.File) !void {
    var bufw = std.io.bufferedWriter(devnull.writer());
    const w = bufw.writer();
    for (0..1 << 20) |i| {
        try w.writeByte(@truncate(i));
    }
    try bufw.flush();
}
// test buffered writer from 0.11.0 which didn't go through AnyWriter
fn testBufferedWriterZig_0_11_0(devnull: std.fs.File) !void {
    var bufw = zig_0_11_0.BufferedWriter(4096, std.fs.File.Writer){ .unbuffered_writer = devnull.writer() };
    const w = bufw.writer();
    for (0..1 << 20) |i| {
        try w.writeByte(@truncate(i));
    }
    try bufw.flush();
}
// test a theoretical writer interface that attempts to be as fast as having no writer
fn testBufferedWriterMeta(devnull: std.fs.File) !void {
    var bufw = BufferedWriterMeta(4096, std.fs.File.Writer){ .unbuffered_writer = devnull.writer() };
    const w = bufw.writer();
    for (0..1 << 20) |i| {
        try w.writeByte(@truncate(i));
    }
    try bufw.flush();
}
// test the performance we get without using any writer
fn testNoWriter(devnull: std.fs.File) !void {
    var buf: [4096]u8 = undefined;
    var idx: usize = 0;
    const w = devnull.writer();
    for (0..1 << 20) |i| {
        buf[idx] = @truncate(i);
        idx += 1;
        if (idx >= buf.len) {
            try w.writeAll(&buf);
            idx = 0;
        }
    }
    try w.writeAll(buf[0..idx]);
}

const mem = std.mem;
const assert = std.debug.assert;

// The old BufferedWriter from 0.11.0
const zig_0_11_0 = struct {
    pub fn GenericWriter(
        comptime Context: type,
        comptime WriteError: type,
        comptime writeFn: fn (context: Context, bytes: []const u8) WriteError!usize,
    ) type {
        return struct {
            context: Context,

            const Self = @This();
            pub const Error = WriteError;

            pub fn write(self: Self, bytes: []const u8) Error!usize {
                return writeFn(self.context, bytes);
            }

            pub fn writeAll(self: Self, bytes: []const u8) Error!void {
                var index: usize = 0;
                while (index != bytes.len) {
                    index += try self.write(bytes[index..]);
                }
            }

            pub fn writeByte(self: Self, byte: u8) Error!void {
                const array = [1]u8{byte};
                return self.writeAll(&array);
            }
        };
    }

    pub fn BufferedWriter(comptime buffer_size: usize, comptime WriterType: type) type {
        return struct {
            unbuffered_writer: WriterType,
            buf: [buffer_size]u8 = undefined,
            end: usize = 0,

            pub const Error = WriterType.Error;
            pub const Writer = GenericWriter(*Self, Error, write);

            const Self = @This();

            pub fn flush(self: *Self) !void {
                try self.unbuffered_writer.writeAll(self.buf[0..self.end]);
                self.end = 0;
            }

            pub fn writer(self: *Self) Writer {
                return .{ .context = self };
            }

            pub fn write(self: *Self, bytes: []const u8) Error!usize {
                if (self.end + bytes.len > self.buf.len) {
                    try self.flush();
                    if (bytes.len > self.buf.len)
                        return self.unbuffered_writer.write(bytes);
                }

                const new_end = self.end + bytes.len;
                @memcpy(self.buf[self.end..new_end], bytes);
                self.end = new_end;
                return bytes.len;
            }
        };
    }
};

// An even more "generic" version of the old GenericWriter that allows the implementation
// to define even more custom implementation other that just a writeFn
pub fn MetaWriter(
    comptime Context: type,
    comptime WriteError: type,
) type {
    return struct {
        context: Context,

        const Self = @This();
        pub const Error = WriteError;

        pub fn writeByte(self: Self, byte: u8) Error!void {
            // An example of how MetaWriter could forward functions to different implementations
            if (comptime @hasDecl(@typeInfo(Context).Pointer.child, "writeByte")) {
                try self.context.writeByte(byte);
            } else {
                const buf = [_]u8{ byte };
                const len = try self.context.write(&buf);
                std.debug.assert(len == 1);
            }
        }

    };
}
pub fn BufferedWriterMeta(comptime buffer_size: usize, comptime WriterType: type) type {
    return struct {
        unbuffered_writer: WriterType,
        buf: [buffer_size]u8 = undefined,
        end: usize = 0,

        pub const Error = WriterType.Error;
        pub const Writer = MetaWriter(*Self, Error);

        const Self = @This();

        pub fn flush(self: *Self) !void {
            try self.unbuffered_writer.writeAll(self.buf[0..self.end]);
            self.end = 0;
        }

        pub fn writer(self: *Self) Writer {
            return .{ .context = self };
        }

        pub fn writeByte(self: *Self, b: u8) Error!void {
            if (self.end == self.buf.len) {
                try self.flush();
            }
            self.buf[self.end] = b;
            self.end += 1;
        }

        pub fn write(self: *Self, bytes: []const u8) Error!usize {
            if (self.end + bytes.len > self.buf.len) {
                try self.flush();
                if (bytes.len > self.buf.len)
                    return self.unbuffered_writer.write(bytes);
            }

            const new_end = self.end + bytes.len;
            @memcpy(self.buf[self.end..new_end], bytes);
            self.end = new_end;
            return bytes.len;
        }
    };
}

```

**Comment by AndrewKraevskii** [comment]
`-Doptimize=ReleaseFast` doen't optimize code when using run command. Correct flag is `-O ReleaseFast`

![image](https://github.com/user-attachments/assets/a7d411fb-ed00-4b86-b1cc-ed3a4fd0156d)
Props to @amp-59 for noticing it

**Comment by marler8997** [comment]
Thanks @AndrewKraevskii...major *facepalm*.  Here's the numbers on my windows machine with `-O ReleaseFast` instead of `-Doptimize=ReleaseFast`.
```
No Writer: 342.3us
BufferedWriter: 471.8us
BufferedWriter(zig 0.11.0): 475.2us
BufferedWriterMeta: 1.102ms
```

**Comment by mlugg** [comment]
OP also accidentally used `-Doptimize` rather than `-O`. I'm editing her post to give the numbers I get for her repro, with `-OReleaseFast`.

**Comment by AndrewKraevskii** [comment]
I tried manually inlining call to write so no AnyWriter present and performance didn't change so devirtualization is indeed not a problem. Also noticed Custom buffering uses AnyWriter in w.writeAll() function anyway)
```zig
pub fn main() !void {
    const devnull = try std.fs.openFileAbsolute("/dev/null", .{ .mode = .write_only });
    var timer = try std.time.Timer.start();
    {
        timer.reset();
        // Write 1MiB to /dev/null, one byte at a time, using std.io.BufferedWriter
        var bufw = std.io.bufferedWriter(devnull.writer());
        const w = bufw.writer();
        for (0..1 << 20) |i| {
            // Change is here vvvvv
            // - try w.writeByte(@truncate(i));
            _ = try w.write(&.{@truncate(i)});
        }
        try bufw.flush();

        std.debug.print("BufferedWriter: {}\n", .{std.fmt.fmtDuration(timer.read())});
    }
    {
        timer.reset();
        // Write 1MiB to /dev/null, one byte at a time, using a custom buffering solution
        var buf: [4096]u8 = undefined;
        var idx: usize = 0;
        const w = devnull.writer();
        for (0..1 << 20) |i| {
            buf[idx] = @truncate(i);
            idx += 1;

            if (idx >= buf.len) {
                try w.writeAll(&buf);
                idx = 0;
            }
        }
        try w.writeAll(buf[0..idx]);

        std.debug.print("Custom buffering: {}\n", .{std.fmt.fmtDuration(timer.read())});
    }
}

const std = @import("std");
```

**Comment by ethernetsellout** [comment]
I went further further and manually removed the `.any()` calls from `GenericWriter`, replacing them with the code from AnyWriter and replacing indirect calls with direct calls to `writeFn`. Again, the performance is unaffected. So this issue is *totally unrelated* to the performance of using AnyWriter, making this all the more peculiar. I think we need better benchmarks to judge this properly, but if this is a geniune performance issue it appears to be a more fundamental issue with Writer and not with dynamic dispatch in the standard library.

> What is the performance impact if devirtualization fails, and is this likely to happen?

There is a severe problem with devirtualization in the case of standard library interfaces that are nested, such as ArenaAllocator or GeneralPurposeAllocator. However, GenericWriter doesn't actually store a child interface like those do, so I doubt devirtualization is ever an issue here.

**Comment by castholm** [comment]
To throw another oddity into the mix, on Windows, if I change `"/dev/null"` to `"\\\\.\\NUL"` and run with `zig run repro.zig -OReleaseFast`, I get the following measurements:
```
BufferedWriter: 4.835ms  
Custom buffering: 412.5us
```
But if I switch out `writeByte` for `writeAll` (still writing a single byte at a time)
```diff
         var bufw = std.io.bufferedWriter(devnull.writer());
         const w = bufw.writer();
         for (0..1 << 20) |i| {
-            try w.writeByte(@truncate(i));
+            try w.writeAll((&@as(u8, @truncate(i)))[0..1]);
         }
         try bufw.flush();
```
and run again, I get a 4x speed increase:
```
BufferedWriter: 1.109ms  
Custom buffering: 411.8us
```
When compiling for `x86_64-linux-gnu` and running with WSL, both the `writeByte` and `writeAll` versions take virtually the same amount of time (1.1ms), so whatever is causing this discrepancy appears to be Windows-specific.

**Comment by nektro** [comment]
related/duplicate https://github.com/ziglang/zig/issues/17985

**Comment by andrewrk** [comment]
in the `wrangle-writer-buffering` branch:
```
BufferedWriter: 444.686us
Custom buffering: 433.985us
```

test case modifications:

```zig
        var buffer: [4096]u8 = undefined;
        var bw: std.io.BufferedWriter = .{
            .buffer = &buffer,
            .unbuffered_writer = devnull.writer(),
        };
        for (0..1 << 20) |i| {
            try bw.writeByte(@truncate(i));
        }
        try bw.flush();

        std.debug.print("BufferedWriter: {D}\n", .{timer.read()});
```

will open a PR shortly

**Comment by andrewrk** [comment]
will resume work on [this branch](https://github.com/ziglang/zig/tree/wrangle-writer-buffering) after 0.14.0 is tagged.

**Comment by andrewrk** [comment]
Updated test. This fills the buffer with random bytes before starting the timer, making the byte values opaque to the optimizer:

```zig
const std = @import("std");

pub fn main() !void {
    const devnull = try std.fs.openFileAbsolute("/dev/null", .{ .mode = .write_only });
    var timer = try std.time.Timer.start();
    var data: [1 << 20]u8 = undefined;
    std.crypto.random.bytes(&data);
    {
        timer.reset();
        // Write 1MiB to /dev/null, one byte at a time, using std.io.BufferedWriter
        var bufw = std.io.bufferedWriter(devnull.deprecatedWriter());
        const w = bufw.writer();
        for (data) |byte| {
            try w.writeByte(byte);
        }
        try bufw.flush();

        std.debug.print("master branch BufferedWriter: {D}\n", .{timer.read()});
    }

    {
        timer.reset();
        // Write 1MiB to /dev/null, one byte at a time, using a custom buffering solution
        var buf: [4096]u8 = undefined;
        var idx: usize = 0;
        const w = devnull.deprecatedWriter();
        for (data) |byte| {
            buf[idx] = byte;
            idx += 1;

            if (idx >= buf.len) {
                try w.writeAll(&buf);
                idx = 0;
            }
        }
        try w.writeAll(buf[0..idx]);

        std.debug.print("master branch custom buffering: {D}\n", .{timer.read()});
    }

    {
        timer.reset();
        var buf: [4096]u8 = undefined;
        var file_writer = devnull.writerStreaming(&buf);
        const w = &file_writer.interface;
        for (data) |byte| {
            try w.writeByte(byte);
        }
        try w.flush();

        std.debug.print("writergate std.io.Writer: {D}\n", .{timer.read()});
    }
}
```

Updated test results (`writergate` branch, `-OReleaseFast`):

```
andy@bark ~/d/z/build-release (writergate)> ./test
master branch BufferedWriter: 813.3us
master branch custom buffering: 818.8us
writergate std.io.Writer: 1.081ms
andy@bark ~/d/z/build-release (writergate)> ./test
master branch BufferedWriter: 808.872us
master branch custom buffering: 830.894us
writergate std.io.Writer: 1.092ms
andy@bark ~/d/z/build-release (writergate)> ./test
master branch BufferedWriter: 849.95us
master branch custom buffering: 815.614us
writergate std.io.Writer: 1.136ms
andy@bark ~/d/z/build-release (writergate)> ./test
master branch BufferedWriter: 751.964us
master branch custom buffering: 806.047us
writergate std.io.Writer: 1.137ms
andy@bark ~/d/z/build-release (writergate)> ./test
master branch BufferedWriter: 766.752us
master branch custom buffering: 817.017us
writergate std.io.Writer: 1.13ms
```

Rust scores about 666us on the same computer:

```rust
$ cat writerand.rs 
use std::fs::File;
use std::io::{BufReader, BufWriter, Read, Write};
use std::time::Instant;
fn main() -> std::io::Result<()> {
    let mut buffer = vec![0u8; 1 << 20];
    let mut urandom = BufReader::new(File::open("/dev/urandom")?);
    urandom.read_exact(&mut buffer)?;
    let dev_null = File::create("/dev/null")?;
    let mut writer = BufWriter::new(dev_null);
    let start = Instant::now();
    for byte in &buffer {
        writer.write_all(&[*byte])?;
    }
    writer.flush()?;
    let duration = start.elapsed();
    println!("Writing 1 MiB to /dev/null, 1 byte at a time, took: {:?}", duration);
    Ok(())
}
```
```
$ rustc --version
rustc 1.86.0 (05f9846f8 2025-03-31) (built from a source tarball)
$ rustc writerand.rs -O
$ ./writerand 
Writing 1 MiB to /dev/null, 1 byte at a time, took: 665.742µs
$ ./writerand 
Writing 1 MiB to /dev/null, 1 byte at a time, took: 663.418µs
$ ./writerand 
Writing 1 MiB to /dev/null, 1 byte at a time, took: 679.399µs
$ ./writerand 
Writing 1 MiB to /dev/null, 1 byte at a time, took: 675.481µs
```

I think the problem here is missed devirtualization. The branch regressed from my earlier testing after switching to use `@fieldParentPtr`. But I'm not going to switch it back. The compiler needs to learn how to devirtualize `@fieldParentPtr`.

Oh, and here are some debug (x86 backend) runs:

```
andy@bark ~/d/z/build-release (writergate)> stage3/bin/zig build-exe test.zig 
andy@bark ~/d/z/build-release (writergate)> ./test
master branch BufferedWriter: 102.389ms
master branch custom buffering: 5.083ms
writergate std.io.Writer: 16.963ms
andy@bark ~/d/z/build-release (writergate)> ./test
master branch BufferedWriter: 101.942ms
master branch custom buffering: 5.474ms
writergate std.io.Writer: 16.882ms
andy@bark ~/d/z/build-release (writergate)> ./test
master branch BufferedWriter: 102.417ms
master branch custom buffering: 4.991ms
writergate std.io.Writer: 16.884ms
```

**Comment by mlugg** [comment]
@andrewrk I was just running the above benchmark to test something, and realised you're being bitten by our array indexing bugs (where `arr[i]` might load the full `arr` onto the stack). If you change the loops to `for (&data)` rather than `for (data)`, you get more encouraging numbers. (In this output I'm also measuring a third data point of pointlessly using `writeInt` for a `u8`; this is relevant to some local perf experiements I'm running.)
```
custom buffering: 320.569us
std.io.Writer.writeByte: 643.382us
std.io.Writer.writeInt: 913.567us
```
Compared to my results for your Rust code:
```
Writing 1 MiB to /dev/null, 1 byte at a time, took: 1.017211ms
```
In other words, it looks like writergate is beating Rust by a not-insigificant margin in this example. Custom buffering is still a good bit faster, but that just gives us something to aim for :^)

Though running your exact cases -- i.e. including the legacy implementation -- shows that writergate is indeed slower than the old `Writer` (I agree with you that this is likely caused by missed devirtualization):
```
master branch BufferedWriter: 500.154us
master branch custom buffering: 314.257us
writergate std.io.Writer: 696.031us
```

**Comment by mlugg** [comment]
I've spent the past few hours looking a bit further into this (essentially playing with reducing various code and seeing how performance characteristics change). The conclusion I have reached is that the absolute differences in times we're seeing are so small that they are actually down to typically-unimportant optimization decisions.

I have a benchmark which optimally takes around 85ms, but changing between different equivalent conditions (e.g. `idx == 4096` vs `idx >= 4096`), or changing a branch hint, or marking a single function as `inline`, or adding/removing a single `switch` prong... can make its runtime vary in the range 85ms--150ms. Running this benchmark using `std.io.Writer` takes around 200ms, which is close enough to optimal that I don't think there's a problem here. I've not been able to make up the difference with any single change; I would get random jumps of improvement between 20-70ms for the kind of arbitrary change I mentioned above, and it was highly inconsistent (the same change could be an improvement, a regression, or a nop, depending on the exact surrounding code).

The benchmark in question is derived from Andrew's: it writes 256 MiB of data to `/dev/null`. So, at 200ms, that's around 1MiB/ms, i.e. around 1GiB/s of throughput. Put differently, each `writeByte` is averaging to a nanosecond. That's approaching a single CPU cycle per byte, so it's quite understandable that small optimization changes can have significant effects on the runtime of the program.

This testing was all done using a buffer size of 4k. I can say with very high confidence that devirtualization is either a) already working for this example, or b) doesn't significantly affect its performance, because manually inlining up to and including the vtable call (so replacing `writeByte` with its definition, and then replacing the `drain` call with a call to the concrete implementation) makes no difference to the performance (in fact, a *lot* of manual inlining and simplification past that has consistently been necessary before I could see any noteworthy performance improvement).

---

To summarize, I think that provided a reasonably large buffer is used (e.g. 4k), the fast concrete buffered path in the new `std.io.Writer` is happily making up for any failed devirtualization. Therefore, my belief is that this issue can be closed. However, it's probably worth waiting to see if any real-world projects struggle with `Writer` performance issues after the writergate upgrade.

**Comment by andrewrk** [comment]
Here's a small data point which does not directly address custom buffering vs std.io.Writer, but it does serve as an argument for considering this issue resolved, along with @mlugg's observation that we are within the realm of optimization stochasticity. 

```c
#include <stdio.h>
#include <string.h>

int main(int argc, char **argv) {
    FILE *file = fopen("/dev/null", "wb");
    char data[1048576];
    memset(data, 0, 1048576);
    for (size_t i = 0; i < 1048576; i += 1) {
        putc(data[i], file);
    }
    fclose(file);
}
```

```zig
const std = @import("std");

pub fn main() !void {
    const devnull = try std.fs.openFileAbsolute("/dev/null", .{ .mode = .write_only });
    defer devnull.close();
    var data: [1048576]u8 = undefined;
    @memset(&data, 0);
    var buf: [1024]u8 = undefined;
    var file_writer = devnull.writerStreaming(&buf);
    const w = &file_writer.interface;
    for (&data) |byte| {
        try w.writeByte(byte);
    }
    try w.flush();
}
```

```
andy@bark ~/s/z/build-release (master)> stage3/bin/zig version
0.15.0-dev.1036+549a466dd
andy@bark ~/s/z/build-release (master)> zig build-exe race.c -target native-native-musl -flto -lc -OReleaseFast --name race-c
andy@bark ~/s/z/build-release (master)> zig build-exe race.zig -OReleaseFast --name race-zig
andy@bark ~/s/z/build-release (master)> poop './race-c' './race-zig'
Benchmark 1 (2434 runs): ./race-c
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          2.02ms ±  152us    1.86ms … 4.14ms        141 ( 6%)        0%
  peak_rss           1.04MB ± 34.5KB     786KB … 1.05MB         43 ( 2%)        0%
  cpu_cycles         6.71M  ±  225K     6.68M  … 12.2M         220 ( 9%)        0%
  instructions       20.0M  ± 1.05      20.0M  … 20.0M           0 ( 0%)        0%
  cache_references   61.3K  ± 7.23K     52.0K  …  204K         503 (21%)        0%
  cache_misses        493   ± 47.4       344   …  956           51 ( 2%)        0%
  branch_misses      4.43K  ± 29.1      4.40K  … 4.53K         279 (11%)        0%
Benchmark 2 (3903 runs): ./race-zig
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          1.26ms ±  122us    1.15ms … 4.40ms        380 (10%)        ⚡- 37.6% ±  0.3%
  peak_rss           1.04MB ± 35.3KB     786KB … 1.05MB         72 ( 2%)          -  0.0% ±  0.2%
  cpu_cycles         2.97M  ± 47.6K     2.95M  … 4.89M          85 ( 2%)        ⚡- 55.8% ±  0.1%
  instructions       10.6M  ± 0.89      10.6M  … 10.6M          24 ( 1%)        ⚡- 46.9% ±  0.0%
  cache_references   54.6K  ± 3.99K     50.4K  … 87.3K         381 (10%)        ⚡- 11.0% ±  0.5%
  cache_misses        360   ± 44.1       203   …  691           72 ( 2%)        ⚡- 27.1% ±  0.5%
  branch_misses      3.39K  ± 17.3      3.36K  … 3.48K         510 (13%)        ⚡- 23.6% ±  0.0%
```
